### PR TITLE
[useAnimatedKeyboard][Android] Fix some keyboard issues

### PR DESCRIPTION
**Pull request by piaskowyk**
## Summary

This pull request addresses a few Android keyboard-related issues:
- It resolves the problem of non-monotonic values during animation. Previously, we were intercepting all keyboard events, but not all of them were relevant to keyboard animation.
- It fixes the issue where the state was incorrect if a transition was canceled.
- It ensures proper respect for auto-hiding navigation bars on Android devices.

## Test plan

Run useAnimatedKeyboard Example on physical Android device.

### CODE REVIEW

File: android/src/main/java/com/swmansion/reanimated/keyboard/KeyboardAnimationCallback.java [hunk 1]
```diff
@@ -31,14 +34,31 @@ public WindowInsetsAnimationCompat.BoundsCompat onStart(
   public WindowInsetsCompat onProgress(
       @NonNull WindowInsetsCompat insets,
       @NonNull List<WindowInsetsAnimationCompat> runningAnimations) {
+    boolean isAnyKeyboardAnimation = false;
```
Comment (tomekzaw) [inline android/src/main/java/com/swmansion/reanimated/keyboard/KeyboardAnimationCallback.java:37]:
```suggestion
    boolean isAnyKeyboardAnimationRunning = false;
```

File: android/src/main/java/com/swmansion/reanimated/keyboard/KeyboardAnimationCallback.java [hunk 2]
```diff
@@ -31,14 +34,31 @@ public WindowInsetsAnimationCompat.BoundsCompat onStart(
...
+        isAnyKeyboardAnimation = true;
+        break;
+      }
+    }
+    if (!isAnyKeyboardAnimation) {
+      return insets;
+    }
     keyboard.updateHeight(insets);
     notifyAboutKeyboardChange.call();
     return insets;
```
Comment (tomekzaw) [inline android/src/main/java/com/swmansion/reanimated/keyboard/KeyboardAnimationCallback.java:49]:
```suggestion

    if (isAnyKeyboardAnimationRunning) {
      keyboard.updateHeight(insets);
      notifyAboutKeyboardChange.call();
    }
    return insets;
```

File: android/src/main/java/com/swmansion/reanimated/keyboard/KeyboardAnimationCallback.java [hunk 3]
```diff
@@ -31,14 +34,31 @@ public WindowInsetsAnimationCompat.BoundsCompat onStart(
...
   }
 
   @Override
   public void onEnd(@NonNull WindowInsetsAnimationCompat animation) {
+    if (!isKeyboardAnimation(animation)) {
+      return;
+    }
     keyboard.onAnimationEnd();
     notifyAboutKeyboardChange.call();
```
Comment (tomekzaw) [inline android/src/main/java/com/swmansion/reanimated/keyboard/KeyboardAnimationCallback.java:58]: This is a matter of taste, I know the code with early returns is more maintainable but this is probably easier to understand. Just a suggestion
```suggestion
    if (isKeyboardAnimation(animation)) {
      keyboard.onAnimationEnd();
      notifyAboutKeyboardChange.call();
    }
```

File: android/src/main/java/com/swmansion/reanimated/keyboard/KeyboardAnimationCallback.java [hunk 4]
```diff
@@ -31,14 +34,31 @@ public WindowInsetsAnimationCompat.BoundsCompat onStart(
...
   }
+
+  private boolean isKeyboardAnimation(@NonNull WindowInsetsAnimationCompat animation) {
```
Comment (tomekzaw) [inline android/src/main/java/com/swmansion/reanimated/keyboard/KeyboardAnimationCallback.java:61]: This method could be static
```suggestion
  private static boolean isKeyboardAnimation(@NonNull WindowInsetsAnimationCompat animation) {
```

**Review by tomekzaw — CHANGES_REQUESTED** [review]

File: android/src/main/java/com/swmansion/reanimated/keyboard/Keyboard.java [hunk 5]
```diff
@@ -20,15 +23,30 @@ public int getHeight() {
...
+    int keyboardHeight = (int) PixelUtil.toDIPFromPixel(Math.max(0, keyboardHeightDip));
+    if (keyboardHeight == 0 && state == KeyboardState.OPEN) {
+      return;
+    }
+    height = keyboardHeight;
   }
 
   public void onAnimationStart() {
-    state = height == 0 ? KeyboardState.OPENING : KeyboardState.CLOSING;
+    if (activeTransitionCounter > 0) {
```
Comment (tjzel) [inline android/src/main/java/com/swmansion/reanimated/keyboard/Keyboard.java:37]: Please check if when switching between two text inputs rapidly you won't get this even re-triggered even though the keyboard shouldn't close.

**Review by tjzel — APPROVED** [review]

File: android/src/main/java/com/swmansion/reanimated/keyboard/Keyboard.java (hunk 5 above)
Comment (piaskowyk) [reply android/src/main/java/com/swmansion/reanimated/keyboard/Keyboard.java:37]: I've checked this and it works 🎉

File: android/src/main/java/com/swmansion/reanimated/keyboard/Keyboard.java [hunk 6]
```diff
@@ -1,34 +1,51 @@
...
     int systemBarBottomInset = insets.getInsets(SYSTEM_BAR_TYPE_MASK).bottom;
-    int keyboardHeightDip = contentBottomInset - systemBarBottomInset;
-    height = (int) PixelUtil.toDIPFromPixel(Math.max(0, keyboardHeightDip));
+    boolean hasNavigationBar = KeyCharacterMap.deviceHasKey(KeyEvent.KEYCODE_HOME);
+    int keyboardHeightDip =
+        hasNavigationBar ? contentBottomInset - systemBarBottomInset : contentBottomInset;
+    int keyboardHeight = (int) PixelUtil.toDIPFromPixel(Math.max(0, keyboardHeightDip));
+    if (keyboardHeight == 0 && mState == KeyboardState.OPEN) {
+      return;
+    }
```
Comment (piaskowyk) [inline android/src/main/java/com/swmansion/reanimated/keyboard/Keyboard.java:32]: TODO: Add comment here

**Pull request closed without merging**